dotcom-rendering/src/model/insert-newsletter-signup.ts

```diff
@@ -0,0 +1,167 @@
+interface PlaceInArticle {
+	index: number;
+	distance: number;
+	afterText: boolean;
+	afterSupporting: boolean;
+}
+
+const isCloseAfterSupportingElement = (
+	index: number,
+	elements: CAPIElement[],
+): boolean => {
+	const elementIsSupporting = (element?: CAPIElement): boolean =>
+		!!(element && 'role' in element && element.role === 'supporting');
+
+	return (
+		elementIsSupporting(elements[index - 1]) ||
+		elementIsSupporting(elements[index - 2])
+	);
+};
+
+const isAfterText = (index: number, elements: CAPIElement[]): boolean =>
+	elements[index - 1]?._type ===
+	'model.dotcomrendering.pageElements.TextBlockElement';
+
+const getPlaces = (
+	targetIndex: number,
+	elements: CAPIElement[],
+): PlaceInArticle[] => {
+	return elements.map((element, index) => ({
+		index,
+		distance: Math.abs(index - targetIndex),
+		afterText: isAfterText(index, elements),
+		afterSupporting: isCloseAfterSupportingElement(index, elements),
+	}));
+};
+
+const placeIsSuitable = (place: PlaceInArticle): boolean =>
+	place.afterText && !place.afterSupporting;
+
+/**
+ * Finds a place within the existing element of the block to insert a
+ * NewsletterSignupBlockElement, within a given distance from a given
+ * target index, prefering a place after TextBlockElement and not within
+ * two places after a supporting element (eg pullquote).
+ *
+ * @param elements the elements in the article block
+ * @param targetIndex the target place to put the NewsletterSignupBlockElement
+ * @param maxDistance how far to move from the targetIndex to find a place after a paragraph
+ * @returns the index insert the NewsletterSignupBlockElement component at
+ */
+export const findInsertIndex = (
+	elements: CAPIElement[],
+	targetIndex: number,
+	maxDistance = 0,
+): number => {
+	const possiblePlaces = getPlaces(targetIndex, elements)
+		.filter(placeIsSuitable)
+		.filter((place) => place.distance <= maxDistance)
```

Comment:
Would it work to fold this check into the `isSuitable` one?

Reply:
(done under another comment)